### DISCUSSION TITLE
Expose PostgREST's `pre-request` config option to enable revoking long-lived tokens

### DISCUSSION THREAD

**Discussion by grschafer** [supabase/supabase#16917] [Contribute to Supabase]
# Feature request

Expose PostgREST's `pre-request` config option to enable revoking long-lived tokens

## Is your feature request related to a problem? Please describe.

This issue is based on this discussion: https://github.com/supabase/supabase/discussions/3205

There is more detail in the discussion linked above, but to summarize my use case:
1. I'd like to give long-lived JWTs to users of my webapp to use as API keys (akin to Stripe tokens or Supabase's `service_role` token)
2. I'd like to let users revoke and regenerate these tokens in my webapp's UI (so they can handle security issues on their end, like leaks or employee offboarding)
3. Supabase currently has no way to apply some security middleware (e.g. to check "is this request's token/claim in my list of revoked tokens?")

There are similar use cases, like this discussion https://github.com/supabase/supabase/discussions/2265 about wanting to revoke a `service_role` JWT (without needing to contact supabase support to do so).

## Describe the solution you'd like

As suggested in [PostgREST's docs on revocation](https://postgrest.org/en/v8.0/tutorials/tut1.html#bonus-topic-immediate-revocation), it would be nice to have a `pre-request` function that can run for all API requests. Maybe this option could be exposed in the Supabase Admin UI as a new API setting? Or, maybe this option could be specified in the database via [PostgREST's In-Database Configuration](https://postgrest.org/en/v8.0/configuration.html#in-database-configuration), though this method seems to require SUPERUSER.

## Describe alternatives you've considered

As described in the github discussion (https://github.com/supabase/supabase/discussions/3205), an alternative would be to add a revocation-check policy to every table, which is not-DRY and is error-prone due to the duplication, as well as bad for performance (if a single request makes many queries, or even a single query across many tables, the revocation check is evaluated repeatedly).

Other alternatives I can think of would be even worse (e.g. putting a revocation-checking proxy service in front of the Supabase API, sending all API requests to some database function that does revocation-checking then dispatches to other tables, e.g. `/rpc/dispatcher?table=profiles&payload=...`).

## Additional context

It looks like y'all are already working on exposing other PostgREST config options, like `db-schema` according to this discussion: https://github.com/supabase/supabase/discussions/1222

Implementing this feature request would hopefully share a lot of the same work and be not-too-onerous as a result! :crossed_fingers:

**Comment by dotlouis** [comment]
Adding to the discussion, this is crucial for my use case.

I'm developing an API and want to provide my users with an API key so they can authorize and authenticate their requests.
I currently have no way to give them a revocable token.

**Comment by psteinroe** [comment]
This would also make life easier when trying to implement multi-tenancy:
- https://github.com/supabase/supabase/discussions/6301
- https://github.com/supabase/supabase/discussions/1148

**Comment by psteinroe** [comment]
I just received answer from the (super awesome) supabase support and they hinted that db-pre-request can already be set!

```sql
ALTER ROLE authenticator SET pgrst.db_pre_request to 'your_custom_function';
NOTIFY pgrst, 'reload config';
```

and to reset

```sql
ALTER ROLE authenticator RESET pgrst.db_pre_request;
NOTIFY pgrst, 'reload config';
```

more details [here](https://postgrest.org/en/stable/configuration.html#in-database-configuration)!

**Comment by ghost** [comment]
@psteinroe I had to run `ALTER ROLE postgres` instead of `ALTER ROLE authenticator` to get this to work. Does this match your experience or what did you do to get `ALTER ROLE authenticator` to work?

**Comment by psteinroe** [comment]
@felixzy-attentec I ended up not using this feature, but you're right - it must be postgres. The code I pasted was copied from the PostgREST documentation and does not reflect the auth setup in the Supabase DB.

**Comment by psteinroe** [comment]
@felixzy-attentec I now need this feature, but the function is never executed. Were you able to configure it for the local environment, too?

**Comment by FelixZY** [comment]
@psteinroe (I'm @felixzy-attentec ) I ended up moving away from supabase as it could not fulfill my needs when building an API. I'm afraid I do not remember my resolution to this specific issue anymore.

**Comment by chipilov** [comment]
@psteinroe In which db schema did you define your custom function?

**Comment by psteinroe** [comment]
@chipilov I tried both a private schema, and public. 

I even created a new project with this sql, and nothing happens.

```sql
create table table_name (
  id bigint generated by default as identity primary key,
  inserted_at timestamp with time zone default timezone('utc'::text, now()) not null,
  updated_at timestamp with time zone default timezone('utc'::text, now()) not null,
  data jsonb,
  name text
);

 CREATE OR REPLACE FUNCTION validate_token ()
   RETURNS void
   LANGUAGE 'plpgsql'
   SECURITY DEFINER
   AS $BODY$
 BEGIN
  raise exception 'test';
 END
 $BODY$;


 ALTER ROLE postgres SET pgrst.db_pre_request TO 'validate_token';
 NOTIFY pgrst, 'reload config';
```

**Comment by chipilov** [comment]
@psteinroe This worked for me by doing ```ALTER ROLE authenticator``` rather than ``` ALTER ROLE postgres```.

A couple of notes:

- I did this against a cloud Supabase instance, haven't tried it locally;
- I set this up a long time ago and while the project that uses this custom function still works (and the function is invoked correctly) there have been many Supabase updates since then so I am not sure if something was changed along the way;

**Comment by psteinroe** [comment]
> This worked for me by doing `ALTER ROLE authenticator` rather than ` ALTER ROLE postgres`.

oh... that worked. never tried it because someone mentioned it has to be `postgres`. Don't know why. Thanks! ❤️
